Warn on line overflow when macro formatting fails

smoelius opened the pull request:
Fixes #5700 by removing these lines: https://github.com/rust-lang/rustfmt/blob/34f9ca28f2f4ae3ac462c6dcd862b513e107e459/src/macros.rs#L138-L141

Re:
> more work will be needed to understand how removing those lines affect macro formatting failures.

I'm not sure what's required here.

All of rustfmt's tests pass with those lines removed.

This change *will* require changes to Clippy. At least the following files are affected:
```sh
find ../rust-clippy -name '*.rs' | while read X; do cargo run --bin rustfmt -- --config-path ../rust-clippy/rustfmt.toml --check "$X" 2>/dev/null || echo "$X"; done
../rust-clippy/tests/ui-internal/check_clippy_version_attribute.rs
../rust-clippy/clippy_lints/src/lib.rs
../rust-clippy/clippy_lints/src/let_underscore.rs
../rust-clippy/clippy_lints/src/utils/internal_lints/lint_without_lint_pass.rs
../rust-clippy/clippy_lints/src/utils/internal_lints.rs
../rust-clippy/clippy_lints/src/utils/mod.rs
../rust-clippy/clippy_lints/src/lib.deprecated.rs
../rust-clippy/clippy_lints/src/derive.rs
../rust-clippy/clippy_lints/src/types/mod.rs
```
Most of the affected lines appear to be `declare_lint_pass!` or `impl_lint_pass!` invocations, e.g.:
https://github.com/rust-lang/rust-clippy/blob/113c704d225c63c1a0eec29cfa9478b7537e7d73/clippy_lints/src/let_underscore.rs#L131
I would be happy to submit the PR to fix those files.

Beyond Clippy, there do appear to be repos that use `error_on_line_overflow=true` "in the wild":
https://grep.app/search?q=error_on_line_overflow%20%2A%3D%20%2Atrue&regexp=true&filter[lang][0]=TOML

But since this an opt-in feature that is [explicitly unstable](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#error_on_line_overflow), maybe it's still okay to change its behavior(?).

r? @ytmimi

ytmimi commented:
@smoelius thanks for the PR. I haven't reviewed this yet but I wanted to quickly address a few things:

> > more work will be needed to understand how removing those lines affect macro formatting failures.
>
> I'm not sure what's required here.

As mentioned in https://github.com/rust-lang/rustfmt/issues/5700#issuecomment-1447586217 PR #3768 originally introduced these changes. We need to make sure we understand why that PR added those lines so we can determine if it's safe to just outright remove those lines.

> This change will require changes to Clippy. At least the following files are affected:

The clippy repo is independant from rustfmt and not under the per-view of the rustfmt team. Feel free to open up a follow up PR there if you'd like but clippy isn't relevant here. 

> Most of the affected lines appear to be `declare_lint_pass!` or `impl_lint_pass!`.

Just want to call out that rustfmt can't format those macros because they don't currently parse, which is related to why adding those lines to the `skip_list` prevented rustfmt from emitting error messages.

> Beyond Clippy, there do appear to be repos that use error_on_line_overflow=true "in the wild":
[https://grep.app/search?q=error_on_line_overflow%20%2A%3D%20%2Atrue&regexp=true&filter[lang][0]=TOML](https://grep.app/search?q=error_on_line_overflow%20%2A%3D%20%2Atrue&regexp=true&filter%5Blang%5D%5B0%5D=TOML)
>
> But since this an opt-in feature that is [explicitly unstable](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#error_on_line_overflow), maybe it's okay to change its behavior(?).

`error_on_line_overflow` is an unstable configuration option, and in my opinion it's fine that we're fixing a bug that would cause rustfmt to emit additional error messages.

ytmimi commented on tests/config/issue_5700.toml:
```diff
@@ -0,0 +1,7 @@
+match_block_trailing_comma = true
+wrap_comments = true
+edition = "2021"
+# error_on_line_overflow = true
```
why is this commented out?

ytmimi commented on tests/rustfmt/main.rs:
```diff
@@ -174,3 +174,20 @@ fn rustfmt_emits_error_on_line_overflow_true() {
+    let args = [
+        "--config-path",
+        "tests/config/issue_5700.toml",
+        "--config",
+        "error_on_line_overflow=true",
```
If we're using a config file already should we just set `error_on_line_overflow=true` there? Although part of me thinks that for clarity we don't need the config file since the only configs that are relevant here are `error_on_line_overflow=true` and `max_width=120`, which we could pass to the test like:

```suggestion
        "--config",
        "error_on_line_overflow=true,max_width=120",
```

ytmimi commented on tests/target/issue_5700.rs:
```diff
@@ -0,0 +1,53 @@
+// rustfmt-config: issue_5700.toml
```
If you decide to remove the `issue_5700.toml` this could change to:

```suggestion
// rustfmt-max_width: 120
```

smoelius commented:
@ytmimi I incorporated your feedback. I agree it is much cleaner without the config file.

I will try to look into #3768.

In the meantime, I have converted the PR to a draft, since, e.g., @calebcartwright wants time to review the issue.

ytmimi commented:
@smoelius Thanks for continuing to help look into this. Based on https://github.com/rust-lang/rustfmt/issues/5700#issuecomment-1452492195, if we decide that the current behavior is accurate, then the simplest solution is to update the docs for [error_on_line_overflow](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#error_on_line_overflow) to mention that we won't alert users when unformattable macros overflow.

smoelius commented:
Do you have a sense for how hard it would be to address the `max_width=100` issue?

When you wrote:
> I don't think we have enough information at the time of calling retrun_macro_parse_failure_fallback to determine how the lines from the input map to lines in the output.

that suggested to me the fix might be non-trivial.

ytmimi commented:
> Do you have a sense for how hard it would be to address the `max_width=100` issue?

I think only the `FmtVisitor` has enough info about the current line number and last byte position of the formatted code. That being said I don't know how involved it would be to properly map lines of text from the input to lines of text in the output.
